## 10.5.5: Generate Commonalities Text Report:

Now we can move on to the final stage of the project - generating text reports to display the results in a human-readable way. Weâ€™ll first display the common proteins between the two genes, then the most-different amino acids between the two.

First, you should print out the common proteins. The function ```display_commonalities(commonalities)``` takes a list of proteins (the proteins held in common between the two genes) and prints each protein in an easy-to-read format, with proteins sorted in alphabetical order (returning None). To make the proteins readable, the format combines the codons of the protein together with dashes between each codon. It also omits the Start and Stop codons (as they are repeated for each protein) and does not include any proteins that are only the Start and Stop codons (in other words, proteins with just two codons).

Given the commonalities produced by the two gene files, this function should print out:

* Ala
* Gly
* Lys
* Ser-Pro-Leu
* Thr

**Note:** to pass the test case, your printed text must match our format exactly!

```python
def display_commonalities(commonalities):
    # remove "Start" and "Stop" codons
    for proteins in commonalities:
        proteins.remove("Start")
        proteins.remove("Stop")

    # arrange proteins in alphabetical order
    commonalities.sort()

    # print out the results
    for proteins in commonalities:
        if len(proteins) > 1:
            separator = "-"
            x = separator.join(proteins)
            print(x)
        else:
            for protein in proteins:
                print(protein)

    return
```